src/ripple/app/tx/impl/Change.cpp

```diff
@@ -189,6 +189,26 @@ Change::applyAmendment()
                 " activated: server blocked.";
             ctx_.app.getOPs ().setAmendmentBlocked ();
         }
+
+        // See if the just enabled amendment retires already-enabled amendments.
+        if (amendment == featureRetire2017Amendments)
+        {
+            // For every retiring amendment, if we find it remove it from
+            // amendmentObject.
+            //
+            // Note that we can't remove the amendments from the AmendmentTable
```

Comment:
I see your perspective.  In a practical sense all validators _should_ be doing this on the same ledger and the outcome should, under ordinary circumstances, be a foregone conclusion.

What I'm attempting to say here is that, in general, we don't assume that all validators move in lock step.  We expect them to produce different ledgers initially and then negotiate a ledger they can all agree on.  Just because the `Change` transaction is applied locally doesn't allow us to assume that the `Change` will be in the validated ledger.  We have to wait for all of the validators to weigh in before we can assume that.

Once we see `featureRetire2017Amendments` in a validated ledger then we're allowed to modify `AmendmentTable`.  But at this point we're not allowed to assume that everyone will agree on `featureRetire2017Amendments`.  So we wait and make the change in `AmendmentTable:: doValidatedLedger`.

If you find the comment confusing I'm happy to remove it or reword it.  Or maybe I'm just confused...